### DISCUSSION TITLE
感谢KKKGO，新版本网关一次成功，另外配置请教下几个小问题

### DISCUSSION THREAD

**Discussion by gukey** [kkkgo/PaoPaoGateWay#9] [问和答]
用了最新版本的 `docker` 固件自定义，确实很容易上手，加上可以替换clash内核，添加配置文件，真的是细节拉满，非常厉害，再次感谢KKKGO的不断努力和付出。

另外有两几个问题请教下：
1、可否添加自定义设置UI面板密码选项，这个是个人需求，如果可以当然最好；
2、最重要的一个问题，关于clash 配置文件生成，我觉得很多人应该都有这个问题，网上教程很多，但是弄个配置文件还是挺麻烦的，目前我是在 `https://v2rayse.com/node-convert/ ` 这个网站转换节点生成的配置文件，也能用，但是想请教你下，你是怎么生成clash配置文件的，原生的配置真的看的人头大，如果可以，能分享下怎么配置的吗，谢谢；
3、关于网络部分配置，麻烦你看看有问题吗：,网关默认就是拨号路由，dns只指定了paopaodns一个IP,目前有个问题是需要代理的设备网关的DNS都需要设置为` 192.168.1.5 ` 才可以正常上网，如果设置 `114.114.114.114`就无法连接网络；
``` sh
ip=192.168.1.6
mask=255.255.255.0
gw=192.168.1.1
dns1=192.168.1.5
```
4、如果安卓手机开启代理软件`NekoBox`，进行部分应用的代理，运行为VPN模式，不连接WIFI可以正常代理，连接WIFI就无法使用，`youtube` 这些都无法访问，单独连接WIFI也能翻墙，但是两个同时都用就不行了，如果连接 WIFI，就需要手动关闭代理软件，然后再重新连接下WIFI，有点繁琐，不知道需要哪里设置一下（同时用是因为`google play`购买了软件，需要不定期自动验证，不然就无法使用一些VIP功能，所以手机正常代理都是一直开的）；

**Comment by kkkgo** [comment]
- 1 UI面板密码是`ppgw.ini`里面的`clash_web_password`设置项。  
- 2 如果你是自己搭的节点，通常就只有一两个，倒也不用经常转换，手写协议也比较简单，你直接参照官方的配置，复制下来自己改就好了，比如这是clash的官方wiki示例配置：https://dreamacro.github.io/clash/configuration/configuration-reference.html   ， 你只需要选取其中的proxies字段填即可：
```yaml
proxies:
  - name: "ss1"
    type: ss
    server: server
    port: 443
    cipher: chacha20-ietf-poly1305
    password: "password"
    udp: true
```
如果你是订阅形式的话，机场应该有提供clash订阅给你，直接用就是了，毕竟现在的机场大多数都是中转，客户端到中转用什么协议并不重要，越简单的协议性能更高，对UDP的支持也更好。如果机场没有提供，你可以尝试使用第三方订阅转换服务（搜索订阅转换），或者使用subconverter这种程序根据自己需求搭一个订阅转换。  
- 3 如果你弄了两个DNS作为主备可以填两个，只有一个就填一个。因为他是基于DNS返回FAKEIP来实现分流的，所以你设置成114他肯定不会走到网关那里去，具体原理可以参见[《FakeIP网关的工作原理》](https://blog.03k.org/post/paopaogateway.html)。
- 4 你需要看看你的代理软件是否有DNS设置选项，把他设置成别的DNS比如1.1.1.1，不然有可能是你的节点或者软件的某个服务器位于境外，然后解析成FAKEIP，跑到网关那里去，而网关又运行着你的节点，变成一个循环。或者你可以把节点或者软件服务相关的域名，添加到PaoPaoDNS的`force_cn_list.txt`，这样他就不会被解析成FAKEIP，详情可以参见PaoPaoDNS的文档：https://github.com/kkkgo/PaoPaoDNS#%E5%8F%82%E6%95%B0%E8%AF%B4%E6%98%8E

**Comment by kkkgo** [comment]
关于3 网络配置部分，我不知道你是说你的设备（手机和电脑）需要设置`192.168.1.5`还是指ppgw需要设置`192.168.1.5`，如果是前者，那肯定是要基于DNS进行分流的；如果是后者，ppgw的DNS仅用于下载`ppgw.ini`和下载订阅，不影响你流量出站，当然，如果你的`ppgw.ini`没有内嵌或者下载失败，或者订阅下载失败，他当然是无法启动正常工作的。

**Comment by kkkgo** [comment]
此外问题4 也很好解决，如果你代理软件是一直挂着的话，那你手动设置你连接WIFI的DNS即可，安卓或者苹果都有这个选项，而且选择只会对你当前连接的WIFI名字单独生效。这个也适用于那些不需要被代理的设备，只需要把DNS改掉他就不会被代理。

**Comment by gukey** [comment]
感谢回复，回来又折腾了一下，我之前是用`ros` 中的`dhcp` 指定网关，吧自己的笔记本和手机都指定到 `paopaogw` 的 `IP`上了，电脑上没有代理软件，就不存在冲突问题，然后我手机上因为持续运行代理，就像你说的，应该是网络变成一个循环了，所以上不了网，现在把ROS网关去了，就正常了，目前使用体验不错；
![](http://forumupload.ru/uploads/001b/b0/1b/8/353011.png)

**Comment by kkkgo** [comment]
额，他设计不是这么用的，如果你把网关直接指向ppgw，得到的全是FAKEIP，会产生很多问题，这个方案目的之一也是减少nat的层数，在ppgw炸了的情况下也不会影响你国内上网，所以他是用静态路由指定FAKEIP网段来用的

**Comment by kkkgo** [comment]
拓扑图可以参见[《FakeIP网关的工作原理》](https://blog.03k.org/post/paopaogateway.html)

**Comment by kkkgo** [comment]
如果你把网关直接指向ppgw相当于全局代理了，他不负责分流的。

**Comment by gukey** [comment]
那我现在如何设置好一些

**Comment by kkkgo** [comment]
你把fakeip的网段`7.0.0.0/8`配置静态路由下一跳到`192.168.1.6`就好了吧，因为DNS会把境外域名都解析到`7.0.0.0/8`这个网段

**Comment by kkkgo** [comment]
当然，如果你在ROS上已经对IP层面进行了分流的话（指把非大陆IP都分流到网关），那custom forward功能也不用开启，静态路由也不用添加了。不过维护IP数据库并不是个好主意，另外BT这些P2P流量也会跑过去，所以fakeip网关的方案是比较灵活的，只会按域名和静态路由来分流，炸了也不影响正常上网

**Comment by gukey** [comment]
三个设备，主路由（192.168.1.1，运行ROS）做DHCP，默认网关为：192.168.1.1，DNS：192.168.1.5；

网关:192.168.1.3  （paopaogw）
DNS:192.168.1.5     （paopaodns）

主路由 192.168.1.1 上 ros 设置了静态路由，无法翻墙，国内能访问，求解：
![](http://forumupload.ru/uploads/001b/b0/1b/8/170346.png)

这是配置PPGW.INI：
```
#paopao-gateway

# mode=socks5|yaml|suburl|free
# default: free
mode=yaml

# Set fakeip's CIDR here
# default: fake_cidr=7.0.0.0/8
fake_cidr=7.0.0.0/8

# Set your trusted DNS here
# default: dns_ip=1.0.0.1
dns_ip=192.168.1.5
# default: dns_port=53
# If used with PaoPaoDNS, you can set the 5304 port
dns_port=5304

# Clash's web dashboard
clash_web_port="80"
clash_web_password="clashpass"

# default：openport=no
# socks+http mixed 1080
openport=no

# default: udp_enable=yes
udp_enable=yes

# default:30
sleeptime=30

# socks5 mode settting
# default: socks5_ip=gatewayIP
socks5_ip="10.10.10.5"
# default: socks5_port="7890"
socks5_port="7890"

# yaml mode settting
# The yaml file in the same directory as the ppgw.ini.
# default: yamlfile=custom.yaml
yamlfile="custom.yaml"

# suburl mode settting
suburl="https://..."
subtime=1d

# yaml and subrul mode setting
# fast_node=check/yes/no
fast_node=yes
test_node_url="https://www.youtube.com/generate_204"
ext_node="Traffic|Expire| GB|Days|Date"
```

这是paopao网关的网络配置：
```
ip=192.168.1.3
mask=255.255.255.0
gw=192.168.1.1
dns1=192.168.1.5
```

**Comment by kkkgo** [comment]
电脑ping一下google看看

**Comment by gukey** [comment]
请求超时

**Comment by kkkgo** [comment]
主要是看解析结果，不是看超时

**Comment by gukey** [comment]
直接不通：
![](http://forumupload.ru/uploads/001b/b0/1b/8/572213.png)

```
   IPv4 地址 . . . . . . . . . . . . : 192.168.1.100(首选)
   子网掩码  . . . . . . . . . . . . : 255.255.255.0
   获得租约的时间  . . . . . . . . . : 2023年6月4日, 星期日 15:24:03
   租约过期的时间  . . . . . . . . . : 2023年6月4日, 星期日 15:55:29
   默认网关. . . . . . . . . . . . . : 192.168.1.1
   DHCP 服务器 . . . . . . . . . . . : 192.168.1.1
   DHCPv6 IAID . . . . . . . . . . . : 141322871
   DHCPv6 客户端 DUID  . . . . . . . : 00-01-00-01-2B-FE-47-49-50-7B-9D-C9-FC-0A
   DNS 服务器  . . . . . . . . . . . : 192.168.1.5
```

**Comment by kkkgo** [comment]
你需要在paopao dns设置custom forward：
```
      - CUSTOM_FORWARD=192.168.1.3:53
      - AUTO_FORWARD=yes
      - AUTO_FORWARD_CHECK=yes
```

**Comment by kkkgo** [comment]
参考说明文档最后一节 与DNS服务器配合完成分流

**Comment by gukey** [comment]
这个具体怎么操作

**Comment by gukey** [comment]
我进docker 设置好像不生效

**Comment by kkkgo** [comment]
如果你是用docker compose来部署容器的话可以直接复制修改，如果是docker命令行部署的话，就是-e指定环境变量，比如`-e CUSTOM_FORWARD=192.168.1.3:53 -e  AUTO_FORWARD=yes`

**Comment by kkkgo** [comment]
或者你用Portainer之类的管理容器的话，点开环境变量一栏就应该可以选择填预设的值。

**Comment by gukey** [comment]
这样子吗？
```
docker run -d \
-e CUSTOM_FORWARD=192.168.1.3:53 -e  AUTO_FORWARD=yes  -e - AUTO_FORWARD_CHECK=yes \
sliamb/paopaodns
```

**Comment by kkkgo** [comment]
是的，不过你这个命令行还要指定一下端口映射或者macvlan（推荐还是docker compose管理比较方便）

**Comment by kkkgo** [comment]
是` -e AUTO_FORWARD_CHECK=yes`不是` -e - AUTO_FORWARD_CHECK=yes`，后者是docker compose的格式

**Comment by gukey** [comment]
这样是不是没生效：
```
localhost:~# docker run -d \
> -e CUSTOM_FORWARD=192.168.1.3:53 -e  AUTO_FORWARD=yes  -e  AUTO_FORWARD_CHECK=yes \
> sliamb/paopaodns
08d904a58be537504c1cc1b7bdeaa7bc1ea3aba672a71eccf995e7ef683d3b0b
localhost:~#
localhost:~#
localhost:~#
localhost:~# docker exec -it paopaodns env
PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
HOSTNAME=0dcc2fbde175
TERM=xterm
CNAUTO=yes
TZ=Asia/Shanghai
UPDATE=weekly
DNS_SERVERNAME=PaoPaoDNS,blog.03k.org
DNSPORT=53
CNFALL=yes
CN_TRACKER=yes
USE_HOSTS=no
IPV6=no
SOCKS5=IP:PORT
SERVER_IP=none
CUSTOM_FORWARD=IP:PORT
AUTO_FORWARD=no
HTTP_FILE=no
HOME=/root
localhost:~#

```

**Comment by kkkgo** [comment]
你上面那个命令没指定容器的名字是`paopaodns `吧，你用`docker ps -a`看看是不是跑着两个容器，你进的是旧容器吧。。。

**Comment by kkkgo** [comment]
docker run -d是新开一个容器，不是修改原来的容器。。

**Comment by gukey** [comment]
不好意思，对docker 不太熟悉，是跑了2个，怎么指定容器修改那？

**Comment by kkkgo** [comment]
docker命令行没有修改的方法，docker compose才可以修改。你可以使用`docker ps -a`查看现存的容器的NAMES一列，然后使用`docker rm -f NAMES`来删除容器，再重建容器：
```shell
#拉取最新的docker镜像  
docker pull sliamb/paopaodns:latest  
#假设你的数据要放在/home/mydata  
docker run -d \
--name paopaodns \
-v /home/mydata:/data \
-e CNAUTO=yes \
-e ..........//写更多环境变量 \
--restart always\
-p 53:53/tcp -p 53:53/udp \
sliamb/paopaodns
```

**Comment by gukey** [comment]
重新部署了，有些网站能用了。youtube 还上不去，需要怎么设置下

**Comment by gukey** [comment]
google 可以上

**Comment by gukey** [comment]
我用了这个配置：
```
#拉取最新的docker镜像  
docker pull sliamb/paopaodns:latest  
#假设你的数据要放在/home/mydata  
docker run -d \
--name paopaodns \
-v /home/mydata:/data \
-e CNAUTO=yes \
-e CUSTOM_FORWARD=192.168.1.3:53  \
-e  AUTO_FORWARD=yes   \
-e  AUTO_FORWARD_CHECK=yes \
--restart always \
-p 53:53/tcp -p 53:53/udp \
sliamb/paopaodns
```

**Comment by gukey** [comment]
在 force_forward_list.txt 添加域名访问正常了，谢谢你帮助；

**Comment by kkkgo** [comment]
那你youtube能ping通不，还有你`ppgw.ini`开了UDP但你节点不支持UDP或者不稳定的话就可能无法访问；  
在 force_forward_list.txt 添加域名访问正常了：那应该是你DNS有缓存吧，没解析到7.0.0.0/8

**Comment by kkkgo** [comment]
此外有些聊天应用是走IP直连的（比如tg），记得把他们的网段加入静态路由。

**Comment by gukey** [comment]
之前一直用旁路由模式做网关翻墙，也用过ros分流，但是效果不好，我的xbox 游戏机启动游戏就无法读取云端存档，没办法放弃了，刚试了一下这次的网络搭建，云端访问正常，能顺利读取存档，体验蛮不错的，再次感谢，要是靠我自己摸索，估计的要花很多时间了，哈哈！

**Comment by gukey** [comment]
> 此外有些聊天应用是走IP直连的（比如tg），记得把他们的网段加入静态路由。

好的。

**Comment by kkkgo** [comment]
如果你是用了auto_forward=yes的话非大陆域名会自动解析到7.0.0.0/8，按道理应该是不用添加域名到force_forward_list的，应该是你电脑本机的DNS缓存吧

**Comment by kkkgo** [comment]
还有，tg的IP段在网上可以搜到，就那么几段。

**Comment by gukey** [comment]
我是把之前的redis_dns.rdb 文件给复制进去了，估计是这的问题啊，之前走的网络不是7.0.0.0

**Comment by gukey** [comment]
是不是现在部署的不需要这个了，让它在慢慢生成；

**Comment by kkkgo** [comment]
`redis_dns.rdb`存的是原始真实数据，不影响的，7.0.0.0/8的结果不会缓存到redis里面的，而是匹配到了就直接转发到网关

**Comment by gukey** [comment]
youtube 解析过，现在注释掉force_forward_list.txt 里面的YouTube.com，也能正常访问；

**Comment by gukey** [comment]
估计还有电脑上的DNS 缓存没清理应该是；

**Comment by gukey** [comment]
另外问下，修改force_forward_list.txt  是需要进docker修改，还是直接在mydata下修改都一样的；

**Comment by kkkgo** [comment]
docker就是挂载映射出来的/data目录，你在外面改和里面改都是同一个文件，修改任何txt列表都是即时重载生效的，如果你是在nas上搭建的甚至可以直接打开smb文件夹双击打开直接记事本改……

**Comment by kkkgo** [comment]
不过你既然用了auto forward，可能需要用到的是force_cn_list.txt，毕竟默认境外域名都走了网关，也许你需要配置的是不走网关的境外域名

**Comment by gukey** [comment]
嗯，后面用到了在加上去；
